src/hardware/mouse/mouse.cpp

```diff
@@ -37,14 +37,260 @@
+	bool gui_takeover = false; // if a GUI requested to take over the mouse
+	bool user_capture = false; // if user requested mouse to be captured
+
+	bool seamless     = false; // if seamless mouse integration is in effect
+	bool drop_events  = true;  // if we should drop mouse events
+
+	bool capture_on_click  = false; // if any button click should capture the mouse
+	bool capture_on_middle = false; // if middle button press should capture the mouse
+	bool release_on_middle = false; // if middle button press should release the mouse
+	bool toggle_on_hotkey  = false; // if hotkey should toggle mouse capture
+
+	bool cursor_outside   = true; // if mouse cursor is outside of drawing area
+	uint16_t cursor_x_abs = 0;    // absolute position from start of drawing area
+	uint16_t cursor_y_abs = 0;
+
+} mouse_gfx;
+
+static void update_cursor_absolute_position(const int32_t x_abs, const int32_t y_abs)
+{
+	mouse_gfx.cursor_outside = false;
+
+	auto calculate = [&](const int32_t absolute,
+	                     const uint16_t clipping,
+	                     const uint16_t resolution) -> uint16_t {
+		assert(resolution > 1u);
+		assert(clipping * 2 < resolution);
+
+		if (absolute < clipping) {
+			mouse_gfx.cursor_outside = true;
+			return 0;
+		} else if (absolute >= resolution - clipping) {
+			mouse_gfx.cursor_outside = true;
+			return static_cast<uint16_t>(resolution - 1);
+		}
+
+		const auto result = absolute - clipping;
+		return static_cast<uint16_t>(result);
+	};
+
+	auto &x = mouse_gfx.cursor_x_abs;
+	auto &y = mouse_gfx.cursor_y_abs;
+
+	x = calculate(x_abs, mouse_gfx.clip_x, mouse_shared.resolution_x);
+	y = calculate(y_abs, mouse_gfx.clip_y, mouse_shared.resolution_y);
+}
+
+static void update_gfx_visibility()
+{
+	// If mouse subsystem not started yet, do nothing
+	if (!mouse_shared.started)
+		return;
+
+	static bool first_time = true;
+
+	// Store internally old settings, to avoid unnecessary GFX calls
+	const auto old_visible = mouse_gfx.visible;
+
+	if (!mouse_gfx.has_focus) { // Window is not in focus
+
+		// No change to cursor visibility
+
+	} else if (mouse_gfx.gui_takeover) { // GUI took over the mouse
+
+		mouse_gfx.visible = true;
+
+	} else { // Window has focus
+
+		// Host cursor should be hidden if any of:
+		// - mouse cursor is captured, for any reason
+		// - seamless integration is in effect
+		// But show it nevertheless if:
+		// - seamless integration is in effect and
+		// - cursor is outside of drawing area
+		mouse_gfx.visible = !(mouse_gfx.captured || mouse_gfx.seamless) ||
+		                    (mouse_gfx.seamless && mouse_gfx.cursor_outside);
+	}
+
+	// Apply calculated settings if changed or if this is the first run
+	if (first_time || old_visible != mouse_gfx.visible)
+		GFX_SetMouseCursor(mouse_gfx.visible);
+
+	// And take a note that this is no longer the first run
+	first_time = false;
+}
+
+static void update_gfx_main()
+{
+	// If mouse subsystem not started yet, do nothing
```

Comment:
(Reviewed up to here; will let you take a pass through and apply high-level feedback, and will continue reviewing after updates).